File: Unity/Assets/PixelPlanets/StellarBreaker/Scripts/Economy/OfflineEarnings.cs
using System;
using StellarBreaker.Core;
using StellarBreaker.Config;

namespace StellarBreaker.Economy
{
    /// <summary>
    /// Offline reward from a timestamp delta: incomePerSecond × min(elapsed, cap) × rate.
    /// </summary>
    public static class OfflineEarnings
    {
        public static double CappedSeconds(long lastUnix, long nowUnix, double capSeconds)
        {
            double elapsed = nowUnix - lastUnix;
            if (elapsed < 0) elapsed = 0;
            return Math.Min(elapsed, capSeconds);
        }

        public static BigNumber Compute(double elapsedSeconds, BigNumber incomePerSecond,
                                        double rate, double capSeconds)
        {
            double secs = Math.Min(Math.Max(0, elapsedSeconds), capSeconds);
            return incomePerSecond * new BigNumber(secs * rate);
        }

        public static BigNumber FromConfig(long lastUnix, long nowUnix,
                                           BigNumber incomePerSecond, BalanceConfig cfg)
        {
            double cap  = cfg.offlineCapHours * 3600.0;
            double secs = CappedSeconds(lastUnix, nowUnix, cap);
            return incomePerSecond * new BigNumber(secs * cfg.offlineRate);
        }
    }
}

File: Telegram/TelegramApp/src/game/economy/OfflineEarnings.ts
// Ported from Assets/PixelPlanets/StellarBreaker/Scripts/Economy/OfflineEarnings.cs
import { BigNumber } from '../core/BigNumber'
import type { BalanceConfig } from '../config/BalanceConfig'

/** Offline reward from a timestamp delta: incomePerSecond * min(elapsed, cap) * rate. */
export function offlineCappedSeconds(lastUnix: number, nowUnix: number, capSeconds: number): number {
  const elapsed = Math.max(0, nowUnix - lastUnix)
  return Math.min(elapsed, capSeconds)
}

export function offlineEarningsFromConfig(
  lastUnix: number,
  nowUnix: number,
  incomePerSecond: BigNumber,
  cfg: BalanceConfig,
  /** Extra hours from the one-time offline cap shop purchase (see MonetizationBoosts). */
  bonusHours = 0,
): BigNumber {
  const cap = (cfg.offlineCapHours + bonusHours) * 3600
  const secs = offlineCappedSeconds(lastUnix, nowUnix, cap)
  return incomePerSecond.mul(new BigNumber(secs * cfg.offlineRate))
}
